R/aggregate_game_stats_def.R

```diff
@@ -0,0 +1,675 @@
+  values_to = "safety_player_id",
+  values_drop_na = TRUE
+  ) %>%
+  dplyr::filter(grepl("_id",desc)) %>%
+  dplyr::mutate(n = 1) %>%
+  tidyr::pivot_wider(
+    names_from = desc,
+    values_from = n,
+    values_fn = sum
+  ) %>%
+  dplyr::select(
+    player_id = safety_player_id,
+    safety = .data$player_id
+  ) %>%
+  dplyr::ungroup()
+
+safety_df_nas <- is.na(safety_df)
+safety_df[safety_df_nas] <- 0
+
+# Penalty stats -----------------------------------------------------------
+
+# get penalty stats
+penalty_df <- data %>%
+  ### the following lines are just to separate defense from offense --> is there a more effitient way???
+  dplyr::filter(!(is.na(penalty_player_id))) %>%
+  dplyr::filter(defteam==penalty_team) %>%
+  ###
+  dplyr::select(
+    tidyselect::contains("penalty_player_")
+  ) %>%
+  tidyr::pivot_longer(cols = c(
+    tidyselect::contains("penalty_player_")
+  ),
+  names_to = "desc",
+  names_prefix = "penalty_",
+  values_to = "penalty_player_id",
+  values_drop_na = TRUE
+  ) %>%
+  dplyr::filter(grepl("_id",desc)) %>%
+  dplyr::mutate(n = 1) %>%
+  tidyr::pivot_wider(
+    names_from = desc,
+    values_from = n,
+    values_fn = sum
+  ) %>%
+  dplyr::select(
+    player_id = penalty_player_id,
+    penalty = .data$player_id
+  ) %>%
+  dplyr::ungroup()
+
+penalty_df_nas <- is.na(penalty_df)
+penalty_df[penalty_df_nas] <- 0
+
+# Touchdown stats -----------------------------------------------------------
+
+# get defensive touchdowns
+touchdown_df <- data %>%
+  dplyr::filter(touchdown==1) %>%
+  dplyr::filter(defteam==td_team) %>%
+  dplyr::group_by(player_id=td_player_id) %>%
+  dplyr::summarise(td = sum(touchdown)) %>%
+  dplyr::ungroup()
+
+touchdown_df_nas <- is.na(touchdown_df)
+touchdown_df[touchdown_df_nas] <- 0
+
+
+# Combine all stats -------------------------------------------------------
+
+# combine all the stats together
+
+player_df <- tackle_df %>%
+  dplyr::full_join(tackle_yds_df, by = c("player_id")) %>%
+  dplyr::full_join(pressure_df, by = c("player_id")) %>%
+  dplyr::full_join(int_df, by = c("player_id")) %>%
+  dplyr::full_join(sack_yds_df, by = c("player_id")) %>%
+  dplyr::full_join(fumble_df_own, by = c("player_id")) %>%
+  dplyr::full_join(fumble_df_opp, by = c("player_id")) %>%
+  dplyr::full_join(fumble_yds_own_df, by = c("player_id")) %>%
+  dplyr::full_join(fumble_yds_opp_df, by = c("player_id")) %>%
+  dplyr::full_join(safety_df, by = c("player_id")) %>%
+  dplyr::full_join(penalty_df, by = c("player_id")) %>%
+  dplyr::full_join(touchdown_df, by = c("player_id")) %>%
+  dplyr::left_join(nflfastR::fast_scraper_roster(2020) %>% dplyr::select(player_id=gsis_id,player_name=full_name,recent_team=team,position)
+                   ) %>%
+  dplyr::select(tidyselect::any_of(c(
+
+    # id information
+    "player_id", "player_name", "recent_team", "position",
+
+    # tackle stats
+    "tkl", "tkl_solo", "tkl_with_assist", "assist", "tfl", "tfl_yards", "forced_fml",
```

Comment:
suggestion: consistency + fully writing out names

"tackles", "tackle_solo", "tackle_assist", (what's tkl_with_assist?), tackle_for_loss, tackle_for_loss_yards, fumbles_forced, sacks, interceptions, passes_defensed, touchdowns, fumbles_forced, fumbles_recovered, fumble_recovery_yards etc

Reply:
"Tackle with assist" + "Tacklo solo" equals to "Tackles". Maybe should remove the "Tackles" sum to avoid misunderstandings. I'll check consistency and names, thx!

Reply:
is a tackle_with_assist the same as an assist? I think that's as confusing to me, they're the same thing in my mind

Reply:
@tanho63 They are different (GSIS IDs 80 and 82). What you think of as an assist is 82. Tackle_with_assist is 80.

Reply:
OK yup I follow that part on the play (some sense of who is the primary player) but for the purposes of aggregating to match this - 
https://www.nfl.com/stats/player-stats/category/tackles/2021/REG/all/defensivecombinetackles/desc

i think they're either duplicate or they add together to one field

Reply:
In the case of that NFL readout, Combined Tackles = Solos + Tackle_with_assist + Assists. So it’s definitely crucial to keep all 3. But you just want an additional sum column of all 3 to check against?

Reply:
as per NFL I don't see the difference between tackle_with_assist + assists and think they should be combined into tackle_assists (and this applies largely to fantasy scoring platforms I've seen also)

Reply:
The NFL sees a difference, though. That's why Solos + Assists don't combine to equal "Combined tackles" in the link you provided. It leaves out Tackle_with_Assist. It's a ghost category. Tackle_with_assist is a Tackle. Assist is an Assist. Tackle_with_assist is almost always classified as a Solo Tackle on fantasy platforms. Because it's a Tackle.

Reply:
should resolve whichever way nfl.com would resolve it

Reply:
That's what I was saying above. The way nfl.com resolves it is by Combined Tackles = Solos + Tackles With Assists + Assists. Assists here will match NFL.com Assists, and Solos will match nfl.com Solos. Just need to add Combined Tackles. But the importance of keeping the 3 distinct in the underlying data is that most Fantasy platforms will show Tackles and Assists, not Solos and Assists (as NFL.com does). And in this case, Tackles = Solos + Tackles with Assist